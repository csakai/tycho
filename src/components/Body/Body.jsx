import React from 'react';
import PropTypes from 'prop-types';
import Constants from '../../constants';

export default class Body extends React.Component {

  static propTypes = {
    radius: PropTypes.number.isRequired,
    atmosphereColor: PropTypes.number.isRequired,
    rotation: PropTypes.object
  };

  render() {
    return (
      <group>
        <mesh rotation={this.props.rotation}>
          <meshPhongMaterial
            specular={Constants.WebGL.SPECULAR_COLOR}
          />
          <sphereGeometry
            widthSegments={Constants.WebGL.SPHERE_SEGMENTS}
            heightSegments={Constants.WebGL.SPHERE_SEGMENTS}
            radius={this.props.radius}
          />
        </mesh>
        <axisHelper size={200} rotation={this.props.rotation} />
      </group>
    );
  }
}
