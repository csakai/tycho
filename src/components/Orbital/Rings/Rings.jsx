import React from 'react';
import PropTypes from 'prop-types';
import { DoubleSide, Euler } from 'three';
import Scale from '../../../utils/Scale';
import MathService from '../../../services/MathService';
import TextureContainer from '../../../containers/TextureContainer';

export default class Rings extends React.Component {

    static propTypes = {
        outerRadius: PropTypes.number.isRequired,
        maps: PropTypes.array.isRequired,
        barycenterTilt: PropTypes.number.isRequired,
        scale: PropTypes.number
    }

    render() {
        const { outerRadius, scale, maps } = this.props;
        const tilt = MathService.toRadians(this.props.barycenterTilt);
        const size = Scale(outerRadius * 2, scale);
        const rotation = new Euler(tilt, 0, 0);

        return (
            <mesh rotation={rotation}>
                <planeGeometry
                    width={size}
                    height={size}
                />
                <TextureContainer
                    transparent={true}
                    side={DoubleSide}
                    textures={maps}
                />
            </mesh>
        );
    }
}
