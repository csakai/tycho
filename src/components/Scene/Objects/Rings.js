import THREE from 'three';
import Scale from 'engine/scale';

export default class Rings extends THREE.Mesh {
  constructor(data) {
    super();

    this.data = data;
    this.render();
  }

  render = () => {
    this.setGeometry();
    this.setMaterial();
  }

  setGeometry = () => {
    const size = Scale(this.data.outerRadius);
    this.geometry = new THREE.PlaneGeometry(size, size, 32);
  }

  setMaterial = () => {
    this.material = new THREE.MeshBasicMaterial({
      color: 0xffff00,
      side: THREE.DoubleSide
    });
  }
}
