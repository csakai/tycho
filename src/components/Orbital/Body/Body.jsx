import React from 'react';
import PropTypes from 'prop-types';
import Constants from '../../../constants';
import Scale from '../../../utils/Scale';
import TextureContainer from '../../../containers/TextureContainer';
import Rings from '../../Orbital/Rings';

export default class Body extends React.Component {

    static propTypes = {
        radius: PropTypes.number.isRequired,
        rings: PropTypes.object,
        rotation: PropTypes.object,
    };

    render() {
        const {rings, radius, scale} = this.props;

        return (
            <group>
                <mesh rotation={this.props.rotation}>
                    <TextureContainer textures={this.props.maps} />
                    <sphereGeometry
                        widthSegments={Constants.WebGL.SPHERE_SEGMENTS}
                        heightSegments={Constants.WebGL.SPHERE_SEGMENTS}
                        radius={Scale(radius, scale)}
                    />
                </mesh>
                {rings && <Rings {...rings} scale={scale} />}
            </group>
        );
    }
}
