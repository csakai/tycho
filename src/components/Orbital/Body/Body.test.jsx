import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {Euler, Object3D} from 'three';
import Body from './';
import Rings from '../Rings';

describe('Body Component', () => {
  let component;
  const radius = 100;
  const rotation = new Euler(1, 1, 1);

  beforeEach(() => {
    component = shallow(<Body
      radius={radius}
      rotation={rotation}
    />);
  });

  it('should render the rings if the rings prop is defined', () => {
    component = shallow(<Body
      radius={radius}
      rotation={rotation}
      rings={{
        outerRadius: 1234,
        barycenterTilt: 20,
        maps: []
      }}
    />);

    expect(component.find(Rings).exists()).toBe(true);
  });

  it('should render the body successfully', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
