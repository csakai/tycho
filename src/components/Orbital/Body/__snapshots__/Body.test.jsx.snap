// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body Component should render the body successfully 1`] = `
<group>
  <mesh
    rotation={
      Euler {
        "_order": "XYZ",
        "_x": 1,
        "_y": 1,
        "_z": 1,
      }
    }
  >
    <TextureContainer />
    <sphereGeometry
      heightSegments={32}
      radius={0.0001}
      widthSegments={32}
    />
  </mesh>
</group>
`;
